frank-doc-frontend/src/app/app.service.ts

```diff
@@ -1,227 +1,106 @@
   }
 
   fullNameToSimpleName(fullName: string): string {
     return fullName.slice(fullName.lastIndexOf('.') + 1);
   }
 
-  getOriginalOrInheritedElement(
-    showInheritance: boolean,
-    element: Element
-  ): Element {
-    const state = this.frankDocStateSource.value;
-    return showInheritance
-      ? this.flattenElements(element) // Merge inherited elements
-      : state.elements[element.fullName]; // Update the element to it's original state
+  filterElementsBySelectedFilters(elements: Element[], selectedFilters: ElementLabels): Element[] {
+    if (this.getSelectedFiltersLength(selectedFilters) === 0) return elements;
+    return elements.filter((element) => {
+      if (!element.labels) return false;
+      for (const labelKey of Object.keys(element.labels)) {
+        if (!selectedFilters[labelKey]) continue;
+        for (const labelItem of element.labels[labelKey]) {
+          if (selectedFilters[labelKey].includes(labelItem)) {
+            return true;
+          }
+        }
+      }
+      return false;
+    });
   }
 
-  flattenElements(element: Element): Element {
-    if (!element.parent) return element;
-
-    const allElements = this.frankDocStateSource.value.elements;
-    const flatElement: Element = { ...element };
-    const parent = allElements[flatElement.parent!];
-
-    //Add separator where attributes inherit from
-    if (parent.attributes && parent.attributes.length > 0) {
-      flatElement.attributes = [...(flatElement.attributes ?? [])]; // attributes was shallowcopied, so we need to spread it in order to use push properly
-      flatElement.attributes.push({ name: '', from: parent });
-    }
-
-    flatElement.attributes =
-      this.copyOf(flatElement.attributes, parent.attributes, 'name') ?? [];
-    flatElement.children =
-      this.copyOf(flatElement.children, parent.children, 'roleName') ?? [];
-    flatElement.parameters =
-      this.copyOf(flatElement.parameters, parent.parameters, 'name') ?? [];
-    flatElement.forwards =
-      this.copyOf(flatElement.forwards, parent.forwards, 'name') ?? [];
-
-    flatElement.parametersDescription ??= parent.parametersDescription;
-    flatElement.parent = parent.parent;
-
-    return this.flattenElements(flatElement);
+  getFirstFilter(labels?: Record<string, string[]>): string {
+    if (!labels) return '-';
+    const labelGroups = Object.keys(labels);
+    if (labelGroups.length === 0) return '-';
+    return labelGroups[0];
   }
 
-  copyOf<T>(
-    baseAttributes: T[] | undefined,
-    mergeAttributes: T[] | undefined,
-    fieldName: keyof T
-  ): T[] | null {
-    if (baseAttributes && !mergeAttributes) return baseAttributes;
-    else if (mergeAttributes && !baseAttributes) return mergeAttributes;
-    else if (!baseAttributes && !mergeAttributes) return null;
+  getFirstLabel(labels?: Record<string, string[]>): string {
+    if (!labels) return '-';
+    const labelGroups = Object.keys(labels);
+    if (labelGroups.length === 0) return '-';
```

Comment:
It might be clearer to what the '-' is if it was a constant.
`if (!labels) return this.NO_RESULTS_CHARACTER;` or something along those lines.